### PR TITLE
add maxps and minps

### PR DESCRIPTION
**Pull request by tathanhdinh**
Hello all,

This PR tries to add `minps` and `maxps` instructions, e.g.

```
0x0        0f 5d 35 57 d6 ff ff        minps xmm6, [rip-0x29a9]
0x7        0f 5d e1                    minps xmm4, xmm1
0xa        0f 5f 85 70 ff ff ff        maxps xmm0, [rbp-0x90]
```

Many thanks for any comment.

### CODE REVIEW

**Comment by pgoodman** [comment]
Can you add tests for these?

File: remill/Arch/X86/Semantics/SSE.cpp [hunk 1]
```diff
@@ -1353,6 +1353,88 @@ IF_AVX(DEF_ISEL(VMAXSD_XMMdq_XMMdq_XMMq) = MAXSD<VV128W, V128, V128>;)
...
+    auto v2 = FExtractV32(src2_vec, i);
+
+    auto min = v1;
+
+    // If either float is a NaN (SNaN or QNaN):
+    if (std::isunordered(v1, v2)) {
+        min = v2;
+    } 
+    // or if both floats are 0.0:
+    else if ((v1 == 0.0) && (v2 == 0.0)) {
```
Comment (pgoodman) [inline remill/Arch/X86/Semantics/SSE.cpp]: Change style to:

```c++
if (...) {
  min = v2;

// or if both floats are 0.0:
} else if ((v1 == 0.0) && (v2 == 0.0)) {
```

File: remill/Arch/X86/Semantics/SSE.cpp [hunk 2]
```diff
@@ -1353,6 +1353,88 @@ IF_AVX(DEF_ISEL(VMAXSD_XMMdq_XMMdq_XMMq) = MAXSD<VV128W, V128, V128>;)
...
+  _Pragma("unroll")
+  for (std::size_t i = 0; i < vec_count; i++) {
+    auto v1 = FExtractV32(dest_vec, i);
+    auto v2 = FExtractV32(src2_vec, i);
+
+    auto min = v1;
+
+    // If either float is a NaN (SNaN or QNaN):
+    if (std::isunordered(v1, v2)) {
+        min = v2;
```
Comment (pgoodman) [inline remill/Arch/X86/Semantics/SSE.cpp]: Two spaces indent, not four.

File: remill/Arch/X86/Semantics/SSE.cpp [hunk 3]
```diff
@@ -1353,6 +1353,88 @@ IF_AVX(DEF_ISEL(VMAXSD_XMMdq_XMMdq_XMMq) = MAXSD<VV128W, V128, V128>;)
...
+    } 
+    // or if both floats are 0.0:
+    else if ((v1 == 0.0) && (v2 == 0.0)) {
+      min = v2;
+    }
+    // or if src2 is greater than src1:
+    else if (v1 >= v2) {
```
Comment (pgoodman) [inline remill/Arch/X86/Semantics/SSE.cpp]: See previous comment.

File: remill/Arch/X86/Semantics/SSE.cpp (hunk 1 above)
Comment (pgoodman) [reply remill/Arch/X86/Semantics/SSE.cpp]: Make sure to add positive and negative zero to test cases. I recommend looking into `__builtin_isless` and variants. Possibly just use `CompareFloats` to implement the comparison to choose the minimum value.

File: remill/Arch/X86/Semantics/SSE.cpp [hunk 4]
```diff
@@ -1353,6 +1353,88 @@ IF_AVX(DEF_ISEL(VMAXSD_XMMdq_XMMdq_XMMq) = MAXSD<VV128W, V128, V128>;)
...
+    auto v1 = FExtractV32(dest_vec, i);
+    auto v2 = FExtractV32(src2_vec, i);
+
+    auto max = v1;
+
+    // If either float is a NaN (SNaN or QNaN):
+    if (std::isunordered(v1, v2)) {
+        max = v2;
+    } 
+    // or if both floats are 0.0:
```
Comment (pgoodman) [inline remill/Arch/X86/Semantics/SSE.cpp]: Style.

File: remill/Arch/X86/Semantics/SSE.cpp [hunk 5]
```diff
@@ -1353,6 +1353,88 @@ IF_AVX(DEF_ISEL(VMAXSD_XMMdq_XMMdq_XMMq) = MAXSD<VV128W, V128, V128>;)
...
+    else if ((v1 == 0.0) && (v2 == 0.0)) {
+      max = v2;
+    }
+    // or if src2 is greater than src1:
```
Comment (pgoodman) [inline remill/Arch/X86/Semantics/SSE.cpp]: Style.

File: remill/Arch/X86/Semantics/SSE.cpp [hunk 6]
```diff
@@ -1353,6 +1353,88 @@ IF_AVX(DEF_ISEL(VMAXSD_XMMdq_XMMdq_XMMq) = MAXSD<VV128W, V128, V128>;)
...
+  }
+  FWriteV32(dst, dest_vec);  // SSE: Writes to XMM, AVX: Zero-extends XMM.
+  return memory;
+}
+
+}
+
+DEF_ISEL(MINPS_XMMps_MEMps) = MINPS<V128W, V128, MV128>;
+DEF_ISEL(MINPS_XMMps_XMMps) = MINPS<V128W, V128, V128>;
+/*
```
Comment (pgoodman) [inline remill/Arch/X86/Semantics/SSE.cpp]: Remove these commented out things now that you've added the ISELs.

File: remill/Arch/X86/Semantics/SSE.cpp [hunk 7]
```diff
@@ -1353,6 +1353,88 @@ IF_AVX(DEF_ISEL(VMAXSD_XMMdq_XMMdq_XMMq) = MAXSD<VV128W, V128, V128>;)
...
+1299 MINPS MINPS_XMMps_MEMps SSE SSE SSE ATTRIBUTES: MXCSR REQUIRES_ALIGNMENT
+1300 MINPS MINPS_XMMps_XMMps SSE SSE SSE ATTRIBUTES: MXCSR REQUIRES_ALIGNMENT
+*/
+
+DEF_ISEL(MAXPS_XMMps_XMMps) = MAXPS<V128W, V128, V128>;
+DEF_ISEL(MAXPS_XMMps_MEMps) = MAXPS<V128W, V128, MV128>;
+/*
```
Comment (pgoodman) [inline remill/Arch/X86/Semantics/SSE.cpp]: Remove these commented out things now that you've added the ISELs.

**Comment by tathanhdinh** [comment]
@pgoodman I will add tests.

**Comment by tathanhdinh** [comment]
Well, tests are failed :). I will come back tomorrow.

**Comment by pgoodman** [comment]
Looks like the main travis failure is that the tests don't actually compile.

**Comment by pgoodman** [comment]
Try `make run-amd64-tests` then `./tests/X86/run-amd64-tests`. The real issue is that the instructions of your test don't actually assemble.

For your own testing purposes, it'll also speed things up if you follow this advice, and only test new things: https://github.com/trailofbits/remill/blob/master/tests/X86/Tests.S#L325

**Comment by tathanhdinh** [comment]
Thanks a lot.
I will do this tomorrow but the problem is that I didn't even succeed to `make build_x86_tests`: it just does nothing (i.e. nothing is compiled), so I didn't know that the tests are compilable or not.

**Comment by pgoodman** [comment]
yes, do `make run-amd64-tests` instead.

**Comment by tathanhdinh** [comment]
Thanks @pgoodman

I've updated the PR, the tests for `min/maxps` are passed now.

NB. I still have some small problems in `make run-amd64-tests` on Fedora, but managed to fix them (I guess that Fedora may be not a supported environment, so never mind).

Please update the document for `make run-amd64-tests` since `make build_x86_tests` might not work anymore (maybe from some [recent commit](#235) about building system?)

**Pull request merged**